nexus/src/app/sagas/test_helpers.rs

```diff
@@ -0,0 +1,300 @@
+    nexus
+        .project_destroy_instance(&opctx, &instance_lookup)
+        .await
+        .expect("Failed to destroy instance");
+}
+
+pub async fn instance_simulate(
+    cptestctx: &ControlPlaneTestContext,
+    instance_id: &Uuid,
+) {
+    info!(&cptestctx.logctx.log, "Poking simulated instance";
+          "instance_id" => %instance_id);
+    let nexus = &cptestctx.server.apictx().nexus;
+    let sa = nexus.instance_sled_by_id(instance_id).await.unwrap();
+    sa.instance_finish_transition(*instance_id).await;
+}
+
+pub async fn instance_simulate_by_name(
+    cptestctx: &ControlPlaneTestContext,
+    name: &str,
+    project_name: &str,
+) {
+    info!(&cptestctx.logctx.log, "Poking simulated instance";
+          "instance_name" => %name,
+          "project_name" => %project_name);
+
+    let nexus = &cptestctx.server.apictx().nexus;
+    let opctx = test_opctx(&cptestctx);
+    let instance_selector =
+        nexus_types::external_api::params::InstanceSelector {
+            project: Some(project_name.to_string().try_into().unwrap()),
+            instance: name.to_string().try_into().unwrap(),
+        };
+
+    let instance_lookup =
+        nexus.instance_lookup(&opctx, instance_selector).unwrap();
+    let (.., instance) = instance_lookup.fetch().await.unwrap();
+    let sa = nexus.instance_sled_by_id(&instance.id()).await.unwrap();
+    sa.instance_finish_transition(instance.id()).await;
+}
+
+/// Tests that a saga `S` functions properly when any of its nodes fails and
+/// causes the saga to unwind by iterating over all saga nodes, creating a new
+/// saga DAG for each node, injecting an error at the chosen node, and verifying
+/// both that the saga failed and that the node at which the failure was
+/// injected was the one that actually caused the saga to fail. This last check
+/// ensures that all possible unwindings are executed.
+///
+/// # Arguments
+///
+/// - `nexus`: A reference to the Nexus that should execute the saga.
+/// - `initial_params`: The parameters to use to construct an initial instance
+///   of saga `S` so that the scaffold can figure out how many nodes are in the
+///   DAG.
+/// - `generate_params`: A callback called at the beginning of each loop
+///   iteration that returns a future that yields the saga parameters to use for
+///   that loop iteration.
+/// - `after_saga`: A callback called after saga execution in each loop
+///   iteration. The caller may use this to check additional post-execution
+///   invariants and to prepare the test for the next loop iteration.
+/// - `log`: A logger to which the scaffold should log messages.
+///
+/// # Panics
+///
+/// This function asserts that each saga it executes (a) starts successfully,
+/// (b) fails, and (c) fails at the specific node at which the function injected
+/// a failure.
+pub async fn action_failure_can_unwind<'a, S, G, P>(
+    nexus: &Arc<Nexus>,
+    initial_params: S::Params,
+    generate_params: G,
```

Comment:
Why bother separating these arguments? Couldn't we just generate `initial_params` using `generate_params`?

(I can see that the snapshot test looks a bit more complex in this area, but most tests appear to simply call `generate_params` to create `initial_params` anyway).

Reply:
This is a somewhat clumsy attempt to require at compile time that the DAGs under test are the same as the DAG that was used to figure out how many nodes to iterate over. I haven't yet found an elegant way to do this; they all have some problems:

1. Create a single DAG in the test and reuse it: this is incorrect if the parameters have to change each time, as in the instance create saga
2. Create parameters for the "query" DAG using the pre-iteration callback: this is incorrect if the callback has side effects, which is allowed (the snapshot test uses them)
3. Create a DAG in the caller and provide it to the harness: this lets the caller pass any `SagaDag` with the underlying saga type erased, though this can be mitigated somewhat by checking that the saga names are the same
4. Make the caller pass an `S::Params` that's independent of the ones generated by the pre-iteration callback: this avoids problem (2) but, as you've noted, looks weird; we're creating fake parameters just to throw the generated DAG away

I ended up going with a different kludge in e5030eb09:

5. Take the DAG created by the first test iteration, count its nodes, and use those for loop bounds; in each iteration, `assert` that the expected node count didn't change

The downside with this method is that we have to do the loop indexing by hand and so are more vulnerable to off-by-one errors than we otherwise would be. But I *think* I've avoided them, and we get the nice benefits that (a) we know the DAG has the same length (if not the same structure) during each iteration, and (b) we can remove a bunch of useless parameter generation from the callers.

Reply:
I like this kludge more! It at least internalizes these details, rather than forcing them on the caller. And IMO the new argument name makes a lot of sense.